src/main/java/com/example/petbuddybackend/service/offer/OfferService.java

```diff
@@ -0,0 +1,278 @@
+    private final OfferRepository offerRepository;
+    private final OfferConfigurationRepository offerConfigurationRepository;
+    private final OfferOptionRepository offerOptionRepository;
+    private final AnimalRepository animalRepository;
+    private final AnimalAttributeRepository animalAttributeRepository;
+    private final AnimalAmenityRepository animalAmenityRepository;
+    private final OfferMapper offerMapper = OfferMapper.INSTANCE;
+    private final OfferConfigurationMapper offerConfigurationMapper = OfferConfigurationMapper.INSTANCE;
+
+    public OfferDTO addOrEditOffer(OfferDTO offer, String caretakerEmail) {
+        Caretaker caretaker = getCaretaker(caretakerEmail);
+
+        Offer modifiyngOffer = getOrCreateCaretakerOffer(caretakerEmail, offer.animal().animalType(),
+                caretaker, offer.description());
+        if(StringUtils.hasText(offer.description())) {
+            modifiyngOffer.setDescription(offer.description());
+        }
+
+        if(CollectionUtil.isNotEmpty(offer.offerConfigurations())) {
+            List<OfferConfiguration> offerConfigurations = createConfigurationsForOffer(offer.offerConfigurations(), modifiyngOffer);
+            if(CollectionUtil.isNotEmpty(modifiyngOffer.getOfferConfigurations())) {
+                modifiyngOffer.getOfferConfigurations().addAll(offerConfigurations);
+            } else {
+                modifiyngOffer.setOfferConfigurations(offerConfigurations);
+            }
+        }
+
+        if(CollectionUtil.isNotEmpty(offer.animalAmenities())) {
+            List<AnimalAmenity> animalAmenities = createAnimalAmenitiesForOffer(offer.animalAmenities(), modifiyngOffer);
+            if(CollectionUtil.isNotEmpty(modifiyngOffer.getAnimalAmenities())) {
+                modifiyngOffer.getAnimalAmenities().addAll(animalAmenities);
+            } else {
+                modifiyngOffer.setAnimalAmenities(animalAmenities);
+            }
+
+        }
+
+        return offerMapper.mapToOfferDTO(offerRepository.save(modifiyngOffer));
+
+    }
+
+    private List<AnimalAmenity> createAnimalAmenitiesForOffer(List<String> animalAmenities, Offer modifiyngOffer) {
+
+        List<AnimalAmenity> newAnimalAmenities = new ArrayList<>();
+        for(String animalAmenity : animalAmenities) {
+            AnimalAmenity newAnimalAmenity = getAnimalAmenity(animalAmenity, modifiyngOffer.getAnimal().getAnimalType());
+            checkDuplicateForAnimalAmenity(
+                    Stream.concat(
+                            Optional.ofNullable(modifiyngOffer.getAnimalAmenities()).orElse(Collections.emptyList()).stream(),
+                            newAnimalAmenities.stream()
+                    ).toList(), newAnimalAmenity
+            );
+            newAnimalAmenities.add(newAnimalAmenity);
+        }
+
+        return newAnimalAmenities;
+
+    }
+
+    private void checkDuplicateForAnimalAmenity(List<AnimalAmenity> oldAnimalAmenities, AnimalAmenity animalAmenity) {
```

Comment:
Skoro nie dopuszczamy duplikatów, to nie lepszym byłoby przechowywanie AnimalAmenity za pomocą Set zamiast List? Wtedy sprawdzenie czy coś istnieje byłoby O(1)

Reply:
racja, set bedzie lepszy, natomiast metode i tak chyba zostawie, zeby zwrocic blad jesli opiekun poda 2 takie same, zeby dac mu o tym znac, ze podal cos zle